docs/decisions/0007-compile-common-theme-for-all-mfes.rst

```diff
@@ -0,0 +1,139 @@
+Paragon/Bootstrap classes, or CSS variables. A separate ADR in `Paragon
+<https://github.com/openedx/paragon/pull/1388>`_ covers the situations where
+this isn't possible yet.
+
+Decision
+********
+
+Common cores styling that is to be applied to all MFEs should be part of a
+separate stylesheet and built separately from the MFE-specific stylesheet.
+
+Both these stylesheets will be included in each MFE, however the common
+stylesheet can be loaded from a common source for all MFEs. This will allow for
+deploying the common theme stylesheet without redeploying MFEs.
+
+The mechanism for loading this common stylesheet will need to be different from
+the mechanism for loading the MFE stylesheet. A JavaScript theme loader will
+contain the location of the SCSS file and will add this stylesheet to the
+document.
+
+The common theme package will be similar to the branding package, but can be
+deployed independently of MFEs. It will produce a CSS stylesheet, and a
+JavaScript loader that will load this stylesheet. When deploying MFEs, they can
+be provided a configuration variable that contains the location where this
+theme loader is deployed.
+
+The frontend-platform repo can host code that simplifies using themes. It can
+check if a theme loader is configured, and if so, dynamically load the theme
+from there. Otherwise, it can continue using the existing mechanism of building
+and deploying the theme for each MFE for backwards-compatibility. It can also
+provide hooks for loading the theme and potentially/eventually also selecting
+a theme.
+
+Consequences
+************
+
+With a single common stylesheet for all MFE, they should load quicker since the
+stylesheet can be cached once for all MFEs. It also makes theme rebuilds much
+quicker, and eliminates the need for rebuilding each MFE for minor change to
+the theme.
+
+For context, on a sample development machine, rebuilding an MFE takes roughly
+one minute, of this, on the same machine, around 3-4 seconds is the time it
+takes to build the theme.
+
+By continuing to support the existing theming mechanism, the current mechanism
+for theming can continue working for those who need it.
+
+Since there is a JavaScript-based loader involved, there is potential for
+flashing of unstyled content during loading, however this can be avoided by
+making the theme loader part of the init process so it runs before the app is
+considered ready.
+
+Another major consequence of this change would be that the CSS code for Paragon
+would be loaded using a different mechanism than the JavaScript code of Paragon.
+The CSS code would be compiled into the theme whereas the JS code will be loaded
+directly. This means that if different versions of Paragon are used for an MFE
+vs the common theme, then there is a chance that an incompatible version of the
+CSS could be loaded for the MFE. This will make it all the more important to
+ensure that version of Paragon used across all MFEs is compatible. This will be
```

Comment:
I am imagining that making paragon compatible only happens one time. Before a mfe uses this compatible version of paragon then it is business as usual. After, they are loading the common css and that is it.

Do you foresee that there could be other moments where compatibility is broken?

Reply:
I'm not sure I understood what you meant by "making paragon compatible only happens one time". 

I think the issue I'm looking at is as follows:
- A bug is encountered in a Paragon component. For example if a label/text for some component is longer than a certain threshold, it will break the UI. 
- The fix for this involves some React code and a change in SCSS. 
- The common compiled theme is built on an older version of Paragon that doesn't include this fix. 
- An MFE is updated to include a newer version of Paragon. 
- The MFE is now using React code, but it's supporting CSS code isn't available in the theme. 

This would need the common theme to be rebuilt as well, using the latest Paragon.

Reply:
I understand now. I thought you were the initial compatibility change, but this issue will come about constantly.